tests/5.2/unified_shared_mem/test_target_VirDestr.cpp

```diff
@@ -0,0 +1,68 @@
+//===--  test_target_VirDestr.cpp-------------------------------------===//
+//
+// OpenMP API Version 5.2
+//
+// Description
+// This test case tests the working of the following aspects:
+// 1) Working of "requires unified_shared_memory"
+// 2) Testing virtual destructor on gpu
+//===----------------------------------------------------------------------===//
+
+
+#include <iostream>
+#include <stdlib.h>
+#include <unistd.h>
+#include <sys/wait.h>
+#include <omp.h>
+#include "ompvv.h"
+
+#define N 1024*512
+
+#pragma omp requires unified_shared_memory
+class base {
+  int *x = nullptr;
+  public:
+    static int StatVar;
+    static void SetStatVal() {
+      StatVar = 999;
+    }
+    base() {
+      x = new int[N];
+      for (int i = 0; i < N; ++i) {
+        x[i] = 10;
+      }
+    }
+    virtual ~base() {
+      delete[] x;
+    }
+};
+
+int base::StatVar = 0;
+class derived: public base {
+  int *y = nullptr;
+  public:
+    derived() {
+      y = new int[N];
+      for (int i = 0; i < N; ++i) {
+        y[i] = 20;
+      }
+    }
+    ~derived() {
+      delete[] y;
+      StatVar = 999;
+    }
+};
+
+int main() {
+  OMPVV_TEST_OFFLOADING;
+  int errors = 0;
+#pragma omp target
+    {
+      derived *d = new derived();
```

Comment:
@lthakur007 , @carlobertolli : I am unclear how is the base function available on the device ? Doesn't the following limitation under target directive apply here:
`A program must not rely on the value of a function address in a target region except for assignments, comparisons to zero and indirect calls.`

Reply:
@spophale I guess the question is: does OpenMP require compilers to support vtable's in device code?
I am not sure if I can get the link to the specification excerpt you indicated: we are effectively talking about an indirect call through the vtable, no? Thanks for the comment!

Reply:
> @spophale I guess the question is: does OpenMP require compilers to support vtable's in device code?

Since OpenMP 5.2, there is in "Restrictions to the **target** construct are as follows:" (in [287:10-12]):

* "[C++] Invoking a virtual member function of an object on a device other than the device on which the object was constructed results in unspecified behavior, unless the object is accessible and was constructed on the host device."

Hence, effectively, OpenMP requires to have something _similar_ to "declare target **indirect**" for the pointer to the vtable and the function points in the vtable.

Disclaimer: I have not really followed the previous discussion nor looked at the example.

Reply:
Given that we are testing the (extent of) support for USM, we can assume that the "object is accessible" condition is met.